python/runtime/feature_derivation/feature_derivation.py

```diff
@@ -0,0 +1,630 @@
+                    %
+                    (field_desc.name, np.prod(field_desc.shape), len(values)))
+
+            field_desc.shape = [len(values)]
+
+    # FIXME(sneaxiy): currently, we only support sparse tensor in CSV format
+    # whose values are 0 or 1. The numeric values in the cell data are the
+    # indices where the values of the sparse tensor are 1. For example, the
+    # cell value "3,5,7" indicates a sparse tensor x, and
+    # x[3] = x[5] = x[7] = 1, and the other values of x are all zeros. Since
+    # the index is always of integer type, we force to set the data type of
+    # sparse tensor in CSV format is "Int". We should remove this constraint
+    # if we will support other data formats in the future.
+    if field_desc.is_sparse:
+        field_desc.dtype = DataType.INT
+
+    field_desc.delimiter = ","
+    for v in values:
+        if field_desc.dtype == DataType.INT:
+            try:
+                int_value = INT64_TYPE(v)
+            except ValueError:
+                field_desc.dtype = DataType.FLOAT
+                field_desc.max_id = 0  # clear the max id
+                continue
+        else:
+            continue
+
+        # INT type, record the maximum id
+        field_desc.max_id = max(field_desc.max_id, int_value)
+
+
+def fill_kv_field_desc(cell, field_desc):
+    """
+    Fill the FieldDesc info by the cell data in the KV format,
+    including shape, etc. of the FieldDesc.
+
+    Args:
+        cell (str): the cell data of the table in the KV format.
+        field_desc (FieldDesc): the FieldDesc object.
+
+    Returns:
+        None.
+    """
+    # split and remove empty string
+    split = [s for s in BLANK_PATTERN.split(cell) if s]
+    max_idx = field_desc.shape[0]
+    for s in split:
+        idx = INT64_TYPE(s.split(':', 2)[0]) + 1
+        if idx > max_idx:
+            max_idx = idx
+
+    field_desc.shape[0] = max_idx
+
+
+def fill_plain_field_desc(cell, field_desc):
+    """
+    Fill the FieldDesc info by the cell data in the PLAIN format,
+    including shape, dtype, vocabulary, etc. of the FieldDesc.
+    This method would try to convert the cell data to be an integer
+    or floating-point number if possible.
+
+    Args:
+        cell (str): the cell data of the table in the PLAIN format.
+        field_desc (FieldDesc): the FieldDesc object.
+
+    Returns:
+        None.
+    """
+    try:
+        int_value = INT64_TYPE(cell)
+    except ValueError:
+        int_value = None
+
+    if int_value is not None:
+        field_desc.shape = [1]
+        return
+
+    try:
+        float_value = float(cell)
+    except ValueError:
+        float_value = None
+
+    if float_value is None:
+        field_desc.dtype = DataType.STRING
+        field_desc.shape = [1]
+        if field_desc.vocabulary is None:
+            field_desc.vocabulary = set()
+        field_desc.vocabulary.add(cell)
```

Comment:
Do we build this vocabulary from the sample data?

Reply:
Yes.